### ISSUE TITLE
Allow CuDNN 7.2 to convert float -> fp16 for tensor core ops

### ISSUE THREAD

**Issue by ThisIsIsaac** [tensorflow/tensorflow#36036]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): Tensorflow using cuDNN 7.2 and above
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
Now, if users want, cuDNN can convert fp32 inputs to fp16 to allow using tensor cores even when inputs are fp32. Before, users had to explicitly convert fp32 to fp16 using AMP. [Link](https://devblogs.nvidia.com/tensor-ops-made-easier-in-cudnn/). This is as simple as setting `CUDNN_TENSOR_OP_MATH_ALLOW_CONVERSION` enum value and passing it to `cudnnSetConvolutionMathType()` functioin.

```cpp
cudnnSetConvolutionMathType(cudnnConvDesc, CUDNN_TENSOR_OP_MATH_ALLOW_CONVERSION)
```

and this is the counterpart for RNN:

```cpp
cudnnSetRNNMatrixMathType(cudnnRnnDesc, CUDNN_TENSOR_OP_MATH_ALLOW_CONVERSION)
```

We should let users chose if they want this option, since it involves converting fp32 to fp16.

**Comment by ThisIsIsaac** [comment]
ping!

This would allow 5 ~ 10x speedup for most convolutions.

**Comment by sanjoy** [comment]
CC @reedwm 

We (TensorFlow team) don't have cycles to actually implement this ourselves, but please feel free to create a PR.

**Comment by reedwm** [comment]
This probably would have been a good idea, but I don't think it's worth doing anymore since [TensorFloat-32](https://github.com/tensorflow/community/pull/247) is superior. TF32 has the advantage in that there's no risk of underflow or overflow compared to fp32, and gradients tend to underflow on the backwards pass in fp16 unless loss scaling is used. Therefore I'm closing this issue.

**Comment by robcaulk** [comment]
Any chance we can reopen this request considering TF32 is only supported by Ampere, where as the `CUDNN_TENSOR_OP_MATH_ALLOW_CONVERSION` is supported by Volta and Turing?

**Comment by ThisIsIsaac** [comment]
If it can help the performance on non-Ampere GPUs, it seems worth awhile. Would love to contribute!

**Comment by sanjoy** [comment]
If we do this, we should unify the API with the TF32 API (or have a good reason for not unifying) https://github.com/tensorflow/community/pull/247.

**Comment by robcaulk** [comment]
> If it can help the performance on non-Ampere GPUs, it seems worth awhile. Would love to contribute!

Do you have a fork/branch where you have tested this? If so, I'd like to compile it and test I on VGG19 inference. 

Question: do you know if the use of this enum value `CUDNN_TENSOR_OP_MATH_ALLOW_CONVERSION` is any different from the environment variable `TF_ENABLE_CUDNN_TENSOR_OP_MATH_FP32` that was experimentally added by nvidia in their [v18.09 package release](https://docs.nvidia.com/deeplearning/frameworks/tensorflow-release-notes/rel_18.09.html#rel_18.09)? The idea of internally casting down to float16 and then back to float32 is described [here](https://docs.nvidia.com/deeplearning/frameworks/tensorflow-user-guide/index.html#tensorcore) for `TF_ENABLE_CUDNN_TENSOR_OP_MATH_FP32`. Meanwhile, the [link you reference](https://devblogs.nvidia.com/tensor-ops-made-easier-in-cudnn/) describes the same process except it discusses automatic padding for NCHW packed data. I can't

**Comment by reedwm** [comment]
@nluehr if we use `CUDNN_TENSOR_OP_MATH_ALLOW_CONVERSION`, users still need to do loss scaling during training, right? I want to confirm that unlike tf32, using tensor cores in this manner can cause overflow due to the lower dynamic range of fp16. If so, I don't think it's worth exposing this functionality, since users still need loss scaling during training and the only further step to use tensor cores would be using mixed precision or quantization

> If we do this, we should unify the API with the TF32 API (or have a good reason for not unifying) tensorflow/community#247.

Unifying will be trickier since the TF32 API has the word `tensor_float_32` in the name. But we can discuss changing this if we decide to allow the use of `CUDNN_TENSOR_OP_MATH_ALLOW_CONVERSION`

> Do you have a fork/branch where you have tested this? If so, I'd like to compile it and test it on VGG19 inference.

Unfortunately I don't think we do.

 > Question: do you know if the use of this enum value CUDNN_TENSOR_OP_MATH_ALLOW_CONVERSION is any different from the environment variable TF_ENABLE_CUDNN_TENSOR_OP_MATH_FP32 that was experimentally added by nvidia in their v18.09 package release? The idea of internally casting down to float16 and then back to float32 is described here for TF_ENABLE_CUDNN_TENSOR_OP_MATH_FP32. Meanwhile, the link you reference describes the same process except it discusses automatic padding for NCHW packed data. That is the only difference I can see without looking through the source code.

@nluehr can you answer this?
